### PR TITLE
Changes to runtimes variables in support of retiring 'whisk/*' runtimes

### PR DESCRIPTION
**Pull request by jonpspri**
Changes to ansible processes for determining and maybe pulling runtime images, as a preliminary to eliminating the hardcoding of the 'whisk' prefix for runtime docker images.

## Description
As per discussion in #3407 , this PR contains changes to ansible processes that will be needed to support the removal of the need for 'whisk/*' images to be in place for builtin runtimes.  Instead,
runtimes _may_ (do we eventually want _should_ or _must_?) be specified in the runtimes.json file.

A pleasant side effect of this change is that all variables (I think) impacting runtimes have now been consolidated into the 'runtimes' YAML object in `ansible/group_vars/all`, and all overrides are tested and defaulted at that point.  There my be some redundant defaulting remaining in `ansible/roles/(controller|invoker)/tasks/deploy.yml`, but I'm trying to manage the number of changes and other maintenance risks that may creep in.

Review and PG requested -- there may be breakage in the IBM Functions build.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [X] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [X] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [X] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [X] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [X] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by jonpspri** [comment]
Yep, I see the Travis failure.  Something funky going on with how the templating of 'whisk.properties' works with my changes.  Need to set up a regression environment to cross-test against.  Gotta go do the day job now, but will likely sort this out during the breaks.

**Comment by jonpspri** [comment]
^^^ Rebased in case a fix to the Kafka heisenbug has been committed.

**Comment by dgrove-oss** [comment]
I thought we decided to move the misc defaults to pureconfig (https://github.com/apache/incubator-openwhisk/blob/8fb3a6fda1700b8cde1de300492b1461a176df15/common/scala/src/main/resources/application.conf#L131-#L137) and removed them from ansible.  Why do they need to move back to ansible (we don't use ansible to deploy on kube, which was why moving the defaults out of ansible was desirable).

**Comment by jonpspri** [comment]
Hi @dgrove-oss, I missed that memo, so forgive me if I’m catching up.  Currently there’s logic in the ansible deploy that pre-fetches the images listed in ‘runtime_manifest.json’ and therefore needs to know the defaults as defined in application.conf.  Unfortunately, ansible doesn’t natively support parsing application.conf (one of several reasons I dislike the pure config approach).  For the time being, that means I do have to keep that separate list of overrides in group_vars/all.  Ideally, I’d want to do away with the pre-loading because Controller/Invoker were smart enough to initialize their own runtimes space, but I don’t think we’re there yet.

**Comment by chetanmeh** [comment]
> Unfortunately, ansible doesn’t natively support parsing application.conf (one of several reasons I dislike the pure config approach)

@jonpspri One possible workaround to address that may be to convert the application.conf to JSON using [typesafe api][2] (see serialization)

```
val config = ConfigFactory.load(); 
val configJSON : String = config.root().render( ConfigRenderOptions.concise() )
```

And then consume that in ansible logic. We can add script to [dev tools][1] which can then be invoked from ansible via gradle command like `gradlew -p tools/dev confToJson`

[1]: https://github.com/apache/incubator-openwhisk/tree/master/tools/dev
[2]: https://static.javadoc.io/com.typesafe/config/1.3.3/com/typesafe/config/Config.html

**Comment by dgrove-oss** [comment]
I'd like to have just one place to get these defaults.  I like @chetanmeh's suggestion of parsing application.conf into ansible so we could make application.conf available to the non-Scala parts of the system. 
@markusthoemmes any thoughts?

**Comment by jonpspri** [comment]
I was thinking about something along @chetanmeh ‘s way of thinking.  Although I’d rather go the other way... a JSON or YAML config file that is used for configuration of the scala code.  Then we’re less likely to run into synchronization problems and edits/overrides are understood.  I expect that’s a non-starter from an effort standpoint.

Can we leave gradle out of this solution?  Ideally in this scenery, one would use ‘shell:’ to invoke an entry point in the invoker jar to dump the JSON to stdout.  That way we avoid all sorts of build out-of-sync issues.

**Comment by chetanmeh** [comment]
>  Ideally in this scenery, one would use ‘shell:’ to invoke an entry point in the invoker jar to dump the JSON to stdout.

Makes sense. We can add a utility main class in common to dump the config. And in ansible just invoke

```
java -cp typesafe-config.jar:openwhisk-common.jar whisk.utils.ConfigDumper /path/to/output.json
```

**Comment by jonpspri** [comment]
I like that approach, _assuming_ we need to continue to pre-pull Docker images in ansible.  My preferred approach would be for Invoker to pre-pull its needed images on startup.  I’m opening the covers on invoker in my next PR, so can we table this conversation until then?  I’m referring #3407 here so we track this conversation.  I’ll also add a TODO to the Ansible files where the suspect variables are defined and used so we don’t build in more dependencies.  We can disappear them over the next few weeks while the JARs are open.  This PRs intention was to do pre-JAR work.  Deal?

**Comment by jonpspri** [comment]
@dgrove-oss Yeah, ansible is doing the pulls based on runtimes.json as well, hence some of this stuff.  If we get invoker to do its own pulls, does that let you cut out a container?  I'm working on the extra Java class now -- probably will have something on Monday.

File: ansible/templates/whisk.properties.j2 [hunk 1]
```diff
@@ -33,7 +33,7 @@ whisk.api.host.name={{ whisk_api_host_name | default(groups['edge'] | first) }}
 whisk.api.localhost.name={{ whisk_api_localhost_name | default(whisk_api_host_name) | default(whisk_api_localhost_name_default) }}
 whisk.api.vanity.subdomain.parts=1
 
-runtimes.manifest={{ runtimesManifest | to_json }}
+runtimes.manifest={{ runtimes.manifest | to_json }}
```
[the remark this answers is not in the document]
Comment (jonpspri) [reply ansible/templates/whisk.properties.j2]: Oh that's scary.  Let me dig through the invoker code -- I'm not sure how much indirection it uses to get the manifest.

[the remark this answers is not in the document]
Comment (rabbah) [reply ansible/templates/whisk.properties.j2]: This file is used for tests only not the invoker. It might still be needed to run tests actually.

**Comment by dgrove-oss** [comment]
it won't help with the KubernetesContainerFactory because in that architecture we don't run an invoker on every node.  (We run a smaller number of invokers as a StatefulSet + an agent (small & slim; written in Go) on every worker node that handles the interactions with docker.

File: ansible/templates/whisk.properties.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonpspri) [reply ansible/templates/whisk.properties.j2]: I'm trying to slowly factor it out of tests.  I'll take a look there.  In the meantime, I guess I'll have to raise an issue so we don't lose track of it.

[the remark this answers is not in the document]
Comment (rabbah) [reply ansible/templates/whisk.properties.j2]: feel free to ignore my comment it's inconsequential for this pr.

**Comment by rabbah** [comment]
We've settled on #3680.

**Pull request closed without merging**